src/ImageSharp/Formats/ImageDecoder{T}.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Six Labors.
+// Licensed under the Six Labors Split License.
+
+using System.IO;
+using System.Threading;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.ImageSharp.Processing;
+
+namespace SixLabors.ImageSharp.Formats
+{
+    /// <summary>
+    /// The base class for all image decoders.
+    /// </summary>
+    /// <typeparam name="T">The type of specialized decoder options.</typeparam>
+    public abstract class ImageDecoder<T> : IImageInfoDetector, IImageDecoder
+        where T : class, ISpecializedDecoderOptions, new()
+    {
+        /// <inheritdoc/>
+        public IImageInfo Identify(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
+        {
+            T specializedOptions = new() { GeneralOptions = options };
+            return this.IdentifySpecialized(specializedOptions, stream, cancellationToken);
+        }
+
+        /// <inheritdoc/>
+        public Image<TPixel> Decode<TPixel>(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
+            where TPixel : unmanaged, IPixel<TPixel>
+        {
+            T specializedOptions = new() { GeneralOptions = options };
+            Image<TPixel> image = this.DecodeSpecialized<TPixel>(specializedOptions, stream, cancellationToken);
+
+            Resize(options, image);
+
+            return image;
+        }
+
+        /// <inheritdoc/>
+        public Image Decode(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
+        {
+            T specializedOptions = new() { GeneralOptions = options };
+            Image image = this.DecodeSpecialized(specializedOptions, stream, cancellationToken);
+
+            Resize(options, image);
+
+            return image;
+        }
+
+        /// <summary>
+        /// Reads the raw image information from the specified stream.
+        /// </summary>
+        /// <param name="options">The specialized decoder options.</param>
+        /// <param name="stream">The <see cref="Stream"/> containing image data.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>The <see cref="IImageInfo"/> object.</returns>
+        /// <exception cref="ImageFormatException">Thrown if the encoded image contains errors.</exception>
+        public abstract IImageInfo IdentifySpecialized(T options, Stream stream, CancellationToken cancellationToken);
```

Comment:
I'm going to see if I can refactor to remove this. There should be no need for specialized options during identification.

Reply:
I think I'd just make this abstract and pass `DecoderOptions`.